In [1]:
import main

INFO:tensorflow:Enabling eager execution
INFO:tensorflow:Enabling v2 tensorshape
INFO:tensorflow:Enabling resource variables
INFO:tensorflow:Enabling tensor equality
INFO:tensorflow:Enabling control flow v2


In [3]:
import pickle
random_board_pairs = []
with open("2021-09-16-mcts-v3-boards12-random.pickle", "rb") as f:
    try:
        while True:
            random_board_pairs.append(pickle.load(f))
    except EOFError:
        pass

In [4]:
for (b,s) in random_board_pairs:
    b.pretty_print_details()
    print(s)

          1111111111
 1234567890123456789
A+++++○++○++++++++++
B++++++++++++++++++○
C+++++○○++++++++○○○+
D+++○++++++○++++++++
E+++++++++++++++++++
F+++○+++++++++++++++
G+++++++++○++++○++++
H+○+++++++++++++++++
I++++++○++++○+○+++++
J++++++++++++++++○++
K++++○+++○++++++++++
L++++++++++++++++++○
M++○++++++++++++++++
N++++++++○++++++++++
O++++++++●++++++++++
Side to move: Left
Moves made: 0
Ball at: [14  8]

0.4679645035795109
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++○+++++++++++
G+++++++++++++++++++
H++●++++++++++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++○+++++++++++++++
N++++++++++++++○++++
O++○++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 2]

0.14394131577885416
          1111111111
 1234567890123456789
A+++++++++○+++++++++
B++++++++++++○++++++
C+++○+++++++++○+++++
D+++○+++++○+++++++++
E+++++++++○++○++++++


In [5]:
len(random_board_pairs)

26

In [1]:
import pickle
random_board_pairs = []
with open("2021-09-16-mcts-v3-boards12-random.pickle", "rb") as f:
    try:
        while True:
            random_board_pairs.append(pickle.load(f))
    except EOFError:
        pass

INFO:tensorflow:Enabling eager execution
INFO:tensorflow:Enabling v2 tensorshape
INFO:tensorflow:Enabling resource variables
INFO:tensorflow:Enabling tensor equality
INFO:tensorflow:Enabling control flow v2


In [2]:
len(random_board_pairs)

38

In [3]:
for (b,s) in random_board_pairs:
    b.pretty_print_details()
    print(s)

          1111111111
 1234567890123456789
A+++++○++○++++++++++
B++++++++++++++++++○
C+++++○○++++++++○○○+
D+++○++++++○++++++++
E+++++++++++++++++++
F+++○+++++++++++++++
G+++++++++○++++○++++
H+○+++++++++++++++++
I++++++○++++○+○+++++
J++++++++++++++++○++
K++++○+++○++++++++++
L++++++++++++++++++○
M++○++++++++++++++++
N++++++++○++++++++++
O++++++++●++++++++++
Side to move: Left
Moves made: 0
Ball at: [14  8]

0.4679645035795109
          1111111111
 1234567890123456789
A+++++++++++++++++++
B+++++++++++++++++++
C+++++++++++++++++++
D+++++++++++++++++++
E+++++++++++++++++++
F+++++++○+++++++++++
G+++++++++++++++++++
H++●++++++++++++++++
I+++++++++++++++++++
J+++++++++++++++++++
K+++++++++++++++++++
L+++++++++++++++++++
M+++○+++++++++++++++
N++++++++++++++○++++
O++○++++++++++++++++
Side to move: Left
Moves made: 0
Ball at: [7 2]

0.14394131577885416
          1111111111
 1234567890123456789
A+++++++++○+++++++++
B++++++++++++○++++++
C+++○+++++++++○+++++
D+++○+++++○+++++++++
E+++++++++○++○++++++


In [4]:
import main

In [5]:
player3 = main.MCTSPlayer3(num_evals= 10000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",)

In [8]:
%%timeit
player3.make_move(random_board_pairs[-1][0])

E NE 
B15
B19
B17
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
B15
B19
B17
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
B15
B19
B17
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
B15
B19
B17
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

In [11]:
player4 = main.MCTSPlayer4(num_evals= 10000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",)

In [12]:
%%timeit
player4.make_move(random_board_pairs[-1][0])

E NE 
B15
B19
B17
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
B15
B19
B17
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
B15
B19
B17
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
B15
B19
B17
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

In [13]:
random_board_pairs[-1][0].pretty_print()

+++○++++++○○+++++++
++++++○+++++++++○++
+○++○++++++++●○++++
+○++++○+++++++○++○+
+++○++++○++++++++++
+++++++++++○++++++○
+++++++○+++++++++++
+++++++++++++++○+++
++++++++○++++○+++++
+++++++++++++++++++
+++○++○++++++++++++
○+++○+++++++○++++++
+++++++++++++++++++
○++++++○+++++++++++
++++++++++++++○++++



In [14]:
bigplayer3 = main.MCTSPlayer3(num_evals= 100000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",)

In [15]:
%%timeit
bigplayer3.make_move(random_board_pairs[-1][0])

E NE 
A17
B19
A16
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
A17
B19
A16
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
A17
B19
A16
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
A17
B19
A16
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

36.9 s ± 1.07 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [16]:
bigplayer4 = main.MCTSPlayer4(num_evals= 100000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",)

In [17]:
%%timeit
bigplayer4.make_move(random_board_pairs[-1][0])

E NE 
A17
B19
A16
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
A17
B19
A16
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
A17
B19
A16
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

E NE 
A17
B19
A16
SE 
A2 0.7016367614269257 1
A3 0.7033960819244385 1
A5 0.7009711265563965 1
A6 0.7016653120517731 1
B2 0.7034583985805511 1
B3 0.7013019621372223 1
B4 0.701871395111084 1
B5 0.7048478126525879 1
B6 0.7049865126609802 1
C3 0.6962560713291168 1
C4 0.697748214006424 1
C6 0.6928581297397614 1
A9 0.7033934891223907 1
A10 0.7024275958538055 1
A13 0.6885223090648651 1
B9 0.7058663666248322 1
B10 0.6995330452919006 1
B11 0.6944043338298798 1
B12 0.7034299671649933 1
B13 0.6513473093509674 1
C9 0.7030385136604309 1
C10 0.7082408368587494 1
C11 0.6672364175319672 1
C12 0.6830804347991943 1
C13 0.6526009142398834 1
A14 0.6884921491146088 1
B14 0.6812886893749237 1
A7 0.7058524191379547 1
A8 0.7067999541759491 1
B8 0.7050804793834686 1
C7 0.703914076089859 1
C8 0.7034156620502472 1
D5 0.7012431025505066 1
D6 0.702261358499527 1
D8 0.7022863030433655 1
D9 0.7053888142108917 1
A15 0.7029431760311127 1
A16 0.695516437292099 1
A17 0.6924314796924591 1
A18 0.7052603960037231 1
A19 0.7

43 s ± 909 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [22]:
import pickle
random_board_pairs = []
with open("2021-09-16-mcts-v3-boards12-random.pickle", "rb") as f:
    try:
        while True:
            random_board_pairs.append(pickle.load(f))
    except EOFError:
        pass

In [23]:
len(random_board_pairs)

58

In [24]:
40000/60

666.6666666666666

In [25]:
40000/72

555.5555555555555

In [27]:
for (b,s) in random_board_pairs:
    if abs(s-0.5)>0.41:
        b.pretty_print_details()
        print(s)
        break

          1111111111
 1234567890123456789
A+++++++++++○+++++++
B++++++++++○+++○++++
C+++++++++++++++++++
D+++++++++++++++++++
E++++++++○++++++++++
F+++++++++++++++++○+
G○+++++++++○++○+++++
H+++++++++++++++○+++
I++++++++++○+++○++++
J++++++++++++○+○+●++
K+++++++++++++++++++
L+++++++++++++++++++
M+++++++++++++++++++
N++○++++○++++++++○++
O++++++++++++++++++○
Side to move: Left
Moves made: 0
Ball at: [ 9 16]

0.9283285775134479


In [28]:
s

0.9283285775134479

In [30]:
player3.num_evals

10000

In [31]:
player3.make_move(b)

K18
K17
S NE 
A10 0.7704465091228485 1
A11 0.7746803462505341 1
A13 0.7741198688745499 1
A14 0.7735776007175446 1
B10 0.7703047096729279 1
B12 0.7745387256145477 1
B13 0.7721050679683685 1
B14 0.774478018283844 1
C10 0.7737216055393219 1
C11 0.7711029797792435 1
C12 0.7729685008525848 1
C13 0.7712236642837524 1
C14 0.7741753906011581 1
A9 0.7739969938993454 1
B9 0.7730975449085236 1
C9 0.7748779356479645 1
D9 0.7739769220352173 1
D10 0.7742010354995728 1
D11 0.7746061980724335 1
D12 0.7734476774930954 1
D13 0.7725093364715576 1
A15 0.7751383036375046 1
A16 0.7737880349159241 1
A17 0.7740332186222076 1
B16 0.7772165685892105 1
B17 0.7688336819410324 1
C15 0.7745046466588974 1
C16 0.7741277813911438 1
C17 0.7731839865446091 1
D14 0.7741751372814178 1
D15 0.7741726785898209 1
D16 0.7733545303344727 1
D17 0.7732698917388916 1
C7 0.7736059725284576 1
C8 0.7743952721357346 1
D7 0.7733374685049057 1
D8 0.7727159261703491 1
E7 0.7727878093719482 1
E8 0.7706953883171082 1
E10 0.7761687934398651

'K18'

In [32]:
smallplayer3 = main.MCTSPlayer3(num_evals= 1000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",)

In [33]:
smallplayer3.make_move(b)

I18
J16
A10
A8
A10 0.7704465091228485 1
A11 0.7746803462505341 1
A13 0.7741198688745499 1
A14 0.7735776007175446 1
B10 0.7703047096729279 1
B12 0.7745387256145477 1
B13 0.7721050679683685 1
B14 0.774478018283844 1
C10 0.7737216055393219 1
C11 0.7711029797792435 1
C12 0.7729685008525848 1
C13 0.7712236642837524 1
C14 0.7741753906011581 1
A9 0.7739969938993454 1
B9 0.7730975449085236 1
C9 0.7748779356479645 1
D9 0.7739769220352173 1
D10 0.7742010354995728 1
D11 0.7746061980724335 1
D12 0.7734476774930954 1
D13 0.7725093364715576 1
A15 0.7751383036375046 1
A16 0.7737880349159241 1
A17 0.7740332186222076 1
B16 0.7772165685892105 1
B17 0.7688336819410324 1
C15 0.7745046466588974 1
C16 0.7741277813911438 1
C17 0.7731839865446091 1
D14 0.7741751372814178 1
D15 0.7741726785898209 1
D16 0.7733545303344727 1
D17 0.7732698917388916 1
C7 0.7736059725284576 1
C8 0.7743952721357346 1
D7 0.7733374685049057 1
D8 0.7727159261703491 1
E7 0.7727878093719482 1
E8 0.7706953883171082 1
E10 0.776168793439865

'K18'

Hmm. It's kind of worrying that in a winning position (score should be ~1), MCTS 1e5 isn't any closer to 1 than MCTS 1e3

In [34]:
for (b,s) in random_board_pairs:
    score = player3.make_move(b)[1]
    if abs(s-score)>0.10:
        b.pretty_print_details()
        print(s)
        break

N10
O8
N12
W 
N8
N6
NE S NE SE 
N12
N14
A4
A2
A4 0.33949899673461914 1
A5 0.34240537881851196 1
A7 0.34287792444229126 1
A8 0.34202659130096436 1
B4 0.3365834355354309 1
B5 0.33976882696151733 1
B6 0.34702157974243164 1
B7 0.3355216979980469 1
B8 0.34453845024108887 1
C4 0.34582656621932983 1
C5 0.3428167700767517 1
C8 0.3396455645561218 1
A10 0.34335505962371826 1
A11 0.34287476539611816 1
B9 0.3421964645385742 1
B10 0.3435705304145813 1
B11 0.34363842010498047 1
C9 0.33614230155944824 1
C10 0.34261858463287354 1
C11 0.3424105644226074 1
A17 0.3430694341659546 1
A18 0.3473427891731262 1
A19 0.342265248298645 1
B17 0.34966790676116943 1
B18 0.3437236547470093 1
C19 0.35249900817871094 1
D17 0.32838916778564453 1
D18 0.3441925644874573 1
D19 0.34646153450012207 1
D5 0.339816153049469 1
D6 0.33788418769836426 1
D7 0.32028788328170776 1
D8 0.3411259651184082 1
E4 0.3375648856163025 1
E5 0.33704888820648193 1
E6 0.33669519424438477 1
E7 0.34350764751434326 1
E8 0.3397431969642639 1
D9 0.34

TypeError: unsupported operand type(s) for -: 'float' and 'str'

In [1]:
import main

INFO:tensorflow:Enabling eager execution
INFO:tensorflow:Enabling v2 tensorshape
INFO:tensorflow:Enabling resource variables
INFO:tensorflow:Enabling tensor equality
INFO:tensorflow:Enabling control flow v2


In [14]:
player3 = main.MCTSPlayer3(num_evals= 10000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",quiet=True)

In [8]:
smallplayer3 = main.MCTSPlayer3(num_evals= 1000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",quiet=True)

In [3]:
import pickle
random_board_pairs = []
with open("2021-09-16-mcts-v3-boards12-random.pickle", "rb") as f:
    try:
        while True:
            random_board_pairs.append(pickle.load(f))
    except EOFError:
        pass

In [9]:
for (b,s) in random_board_pairs:
    score = smallplayer3.make_move(b,return_all=True)[1]
    if abs(s-score)>0.10:
        b.pretty_print_details()
        print(s)
        break

          1111111111
 1234567890123456789
A++++++++○++++○++○++
B++++++++++++++○++++
C++++++++++++++++○++
D+++++++++++++++++++
E+++++++++++++++○+++
F+++++++++++++++++○+
G+++++++++++++++○++○
H+++++++++++++++++++
I++++++++○+++○++++++
J++++++++++○●++++○++
K+++++++++++++++++++
L+++++++++++++++○+++
M++++++++++++++○+++○
N+++++++++++○○++++++
O++++++++++++++++○++
Side to move: Right
Moves made: 0
Ball at: [ 9 11]

0.4344205332108033


In [10]:
loudsmallplayer3 = main.MCTSPlayer3(num_evals= 1000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",quiet=False)

In [19]:
loudplayer3 = main.MCTSPlayer3(num_evals= 1000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model")

In [11]:
loudsmallplayer3.make_move(b)

W NW 
H9
G7
A7
A5
A7 0.3321290612220764 1
A8 0.3378492593765259 1
A10 0.33802664279937744 1
A11 0.34167248010635376 1
B7 0.3374316692352295 1
B8 0.3377441167831421 1
B9 0.34007740020751953 1
B10 0.3378256559371948 1
B11 0.33799242973327637 1
C7 0.3373812437057495 1
C8 0.3375272750854492 1
C9 0.3392679691314697 1
C10 0.3391542434692383 1
C11 0.3374497890472412 1
A12 0.3375362157821655 1
A13 0.3378230333328247 1
A15 0.3373457193374634 1
A16 0.3357802629470825 1
B12 0.33847367763519287 1
B13 0.3486534357070923 1
B14 0.33789682388305664 1
B16 0.33628612756729126 1
C12 0.33763551712036133 1
C13 0.34191715717315674 1
C14 0.3363267779350281 1
C15 0.325583815574646 1
C16 0.3376007676124573 1
A18 0.33899617195129395 1
A19 0.33829283714294434 1
B17 0.33914196491241455 1
B18 0.3388035297393799 1
B19 0.3374888300895691 1
C18 0.33748364448547363 1
C19 0.3345072269439697 1
D13 0.3374119997024536 1
D14 0.3319048285484314 1
D15 0.32852405309677124 1
D16 0.334314227104187 1
D17 0.33350467681884766 1
D1

'W NW '

In [13]:
for (b,s) in random_board_pairs:
    score = smallplayer3.make_move(b,return_all=True)[1]
    print(abs(s-score))

0.01807484369791912
0.08352879769365087
0.083568618673108
0.07326546818296487
0.017006186306053883
0.041089525961744644
0.0008240982492282445
0.1401287458318493
0.02428539594014484
0.11295586576064431
0.013702942846129829
0.0749963764088376
0.01177649953866633
0.04684434424874917
0.05447854505201272
0.02481459419460652
0.014183774388822387
0.021615294086802295
0.05355238914489746
0.04627172289223502
0.006814329742508757
0.09354641769319921
0.023659474617586773
0.054317739293202605
0.0003231080673143927
0.07814854016276696
0.024316876080998195
0.08823891803842032
0.0722967124935896
0.009345627625783282
0.0
0.016586770529270933
0.03428220244113511
0.02134209508831436
0.042088339487329596
0.17483799898253372
0.1110963277092582
0.038516574870539055
0.009011064560796367
0.028810185311686
0.029582555772680208
0.07908395784241812
0.02808814022130951
0.013179240920163982
0.06706506467163563
0.04447570456927563
0.03196834087470646
0.011071968646276575
0.056506387622209886
0.07240963943542977
0.

In [15]:
for (b,s) in random_board_pairs:
    score = player3.make_move(b,return_all=True)[1]
    print(abs(s-score))

0.002413340941195552
0.019234687021610464
0.020397207189604694
0.021224729773749118
0.05027574769496712
0.0153715186629374
0.01220246532391478
0.020676511035435274
0.007842354914721295
0.010925355411711246
0.05353893921116426
0.03338360797646922
0.03615552710304315
0.022086715319000194
0.05844907654639947
0.03266355970495044
0.02905400653978936
0.01895399210096771
0.07937377810478208
0.006542942304720745
0.00907857087166869
0.01304056162578393
0.019786118185864066
0.008480494584715281
0.009166323983507363
0.025664494977835295
0.013859700215490212
0.010529793304415946
0.05642761475110514
0.01613235394159951
0.0
0.010625774425372048
0.0020184832748165182
0.05009616052245233
0.048161013304212075
0.18263807565230805
0.02982125141431946
0.00638622893911911
0.038845218055788444
0.02034068174386683
0.013767923407891236
0.043209731120329664
0.012431249547995848
0.017986910840584014
0.014258521611609704
0.001295047266935745
0.057711713825941535
4.257474706115261e-09
0.03684063575860025
0.003036

Hmm. So MCTS 1e4 seems to have fewer big divergences from MCTS 1e5's score than MCTS 1e3 does.

We are sure that MCTS 1e5 is a better player than 1e4, right?

In [17]:
(b,s) = random_board_pairs[-23]
score = player3.make_move(b,return_all=True)[1]
print(abs(s-score))

0.18263807565230805


In [18]:
b.pretty_print_details()

          1111111111
 1234567890123456789
A+++++++++++++○+++++
B++++++++++++++++○++
C++++++++++++++++○++
D+++++++++++++++++○+
E+++○+++++++++++++++
F+++++++++++++++++++
G+++++++++++++++++++
H+++++++○+++++++++++
I+++++++○+++++++++++
J+++++++++○+++++++++
K++++○+++++++++○+++○
L+++++++++++++++++++
M+++++○+++++++++++++
N+++++++++++++++++++
O○+○++●+++++++++++++
Side to move: Right
Moves made: 0
Ball at: [14  5]



In [20]:
loudplayer3.make_move(b)

O5
N7
W W W 
A12 0.6399626433849335 1
A13 0.6410426199436188 1
A15 0.638934999704361 1
A16 0.6400730609893799 1
B12 0.6410613358020782 1
B13 0.6410878002643585 1
B14 0.6426113545894623 1
B15 0.6411377787590027 1
B16 0.6400100290775299 1
C12 0.6412846148014069 1
C13 0.6421559751033783 1
C14 0.6414259076118469 1
C15 0.6368692815303802 1
C16 0.6396138966083527 1
A17 0.6385335028171539 1
A18 0.6370193064212799 1
A19 0.6402798891067505 1
B18 0.6298745572566986 1
B19 0.6322804987430573 1
C18 0.6412762701511383 1
C19 0.6359133422374725 1
D15 0.6339183151721954 1
D16 0.6406248509883881 1
D17 0.6387095749378204 1
D19 0.6403546035289764 1
E15 0.6415720283985138 1
E16 0.6401445269584656 1
E17 0.6401967108249664 1
E18 0.6424740254878998 1
E19 0.6361072957515717 1
F16 0.641169399023056 1
F17 0.6413349211215973 1
F18 0.641479879617691 1
F19 0.6408810317516327 1
C2 0.6403845846652985 1
C3 0.640386551618576 1
C4 0.6418643891811371 1
C5 0.6423453986644745 1
C6 0.6401902437210083 1
D2 0.6405551731586456

'O5'

In [25]:
loudbigplayer3 = main.MCTSPlayer3(num_evals= 100000, curiosity = 0.001, name="saved_model/2021-08-23-v3-model",quiet=False)

In [26]:
loudbigplayer3.make_move(b)

O5
W W 
W 
A12 0.6399626433849335 1
A13 0.6410426199436188 1
A15 0.638934999704361 1
A16 0.6400730609893799 1
B12 0.6410613358020782 1
B13 0.6410878002643585 1
B14 0.6426113545894623 1
B15 0.6411377787590027 1
B16 0.6400100290775299 1
C12 0.6412846148014069 1
C13 0.6421559751033783 1
C14 0.6414259076118469 1
C15 0.6368692815303802 1
C16 0.6396138966083527 1
A17 0.6385335028171539 1
A18 0.6370193064212799 1
A19 0.6402798891067505 1
B18 0.6298745572566986 1
B19 0.6322804987430573 1
C18 0.6412762701511383 1
C19 0.6359133422374725 1
D15 0.6339183151721954 1
D16 0.6406248509883881 1
D17 0.6387095749378204 1
D19 0.6403546035289764 1
E15 0.6415720283985138 1
E16 0.6401445269584656 1
E17 0.6401967108249664 1
E18 0.6424740254878998 1
E19 0.6361072957515717 1
F16 0.641169399023056 1
F17 0.6413349211215973 1
F18 0.641479879617691 1
F19 0.6408810317516327 1
C2 0.6403845846652985 1
C3 0.640386551618576 1
C4 0.6418643891811371 1
C5 0.6423453986644745 1
C6 0.6401902437210083 1
D2 0.6405551731586456 1

'O5'

In [21]:
s

0.9990958037797796

Ah, in this case, 1e5 is correctly sure (0.999) that it's going to win, whereas 1e4, despite making the same winning move, is only at 0.82. 

So this reassures me that sometimes 1e5 does have a better score than 1e4. And hopefully collecting the expensive 1e5 data will be worth it

In [27]:
import pickle
random_board_pairs = []
with open("2021-09-18-mcts-v3-boards13-random.pickle", "rb") as f:
    try:
        while True:
            random_board_pairs.append(pickle.load(f))
    except EOFError:
        pass

In [28]:
len(random_board_pairs)

12